pkg/endpoint/api/endpoint_deletion_queue.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		dq.logger.Warn("Failed to lock queued deletion directory, proceeding anyways. This may cause CNI deletions to be missed.",
+			logfields.Path, defaults.DeleteQueueLockfile,
+			logfields.Error, err,
+		)
+		return nil
+	}
+
+	if err = dq.lf.Lock(ctx, true); err != nil {
+		dq.logger.Warn("Failed to lock queued deletion directory, proceeding anyways. This may cause CNI deletions to be missed.",
+			logfields.Path, defaults.DeleteQueueLockfile,
+			logfields.Error, err,
+		)
+		dq.lf.Close()
+		dq.lf = nil
+	}
+	return nil
+}
+
+// processQueuedDeletes is the agent-side of the identity deletion queue.
+// The CNI plugin queues deletions when the agent is down, because
+// containerd / crio expect CNI DEL to always succeed. It does so
+// by writing a file to /run/cilium/deleteQueue with the endpoint ID
+// to delete.
+//
+// On startup, we will grab a lockfile in this directory, then process
+// all deletions. Then, we start up the agent server, then drop the lock.
+// Any CNI processes waiting in that period of time will, after getting
+// the lock.
+func (dq *DeletionQueue) processQueuedDeletes(ctx context.Context) error {
+	files, err := filepath.Glob(defaults.DeleteQueueDir + "/*.delete")
+	if err != nil {
+		dq.logger.Error("Failed to list queued CNI deletion requests. CNI deletions may be missed.",
+			logfields.Path, defaults.DeleteQueueDir,
+			logfields.Error, err,
+		)
+	}
+
+	dq.logger.Info("Processing queued deletion requests", logfields.LenEndpoints, len(files))
+
+	for _, file := range files {
+		select {
+		case <-ctx.Done():
+			// stop processing on context cancellation
+			return nil
+		default:
+		}
+
+		err = dq.processQueuedDeleteEntryLocked(file)
+		if err != nil {
+			dq.logger.Error("Failed to read queued CNI deletion entry. Endpoint will not be deleted.",
+				logfields.Path, file,
+				logfields.Error, err,
+			)
+		}
+
+		if err := os.Remove(file); err != nil {
+			dq.logger.Error("Failed to remove queued CNI deletion entry, but deletion was successful.",
+				logfields.Path, file,
+				logfields.Error, err,
+			)
+		}
+	}
+
+	return nil
+}
+
+// unlockAfterAPIServer registers a start hook that runs after API server
+// has started and the deletion queue has been drained to unlock the
+// delete queue and thus allow CNI plugin to proceed.
+func unlockAfterAPIServer(jobGroup job.Group, _ *server.Server, dq *DeletionQueue) {
+	jobGroup.Add(job.OneShot("unlock-lockfile", func(ctx context.Context, health cell.Health) error {
+		// Explicitly wait until deletion queue finished processing or job context is cancelled
+		select {
+		case <-ctx.Done():
+			// continue and unlock
+		case <-dq.processed:
+			// continue and unlock
+		}
+
+		if dq.lf != nil {
+			unlockErr := dq.lf.Unlock()
+			closeErr := dq.lf.Close()
+
+			if unlockErr != nil || closeErr != nil {
+				return fmt.Errorf("failed to unlock deletion queue lock file: %w", errors.Join(unlockErr, closeErr))
+			}
+		}
+
+		return nil
+	}))
+}
+
```

Comment:
Why not put this logic into `dq.Process()`? It's an unusual pattern to have one goroutine grab a lock and then delegate the release of the lock to another goroutine. I think that strictly speaking given the synchronization between the goroutines with the channel it's probably technically correct, but this seems ripe for abuse if the logic is ever extended. As a general rule locks should be held in a single goroutine.